core/src/main/java/org/apache/gravitino/stats/storage/PartitionStatisticStorage.java

```diff
@@ -0,0 +1,117 @@
+   * @return a map where the key is the partition name and the value is another map * containing
+   *     statistic names and their corresponding values.
+   */
+  Map<String, Map<String, StatisticValue<?>>> listStatistics(
+      MetadataObject metadataObject, PartitionRange partitionRange);
+
+  /**
+   * Lists statistics for a given metadata object and specific partition names. This interface may
+   * be used in the future. The upper logic layer won't call this method now.
+   *
+   * @param metadataObject the metadata object for which statistics are being listed
+   * @param partitionNames a list of partition names for which statistics are being listed
+   * @return a map where the key is the partition name and the value is another map.
+   */
+  default Map<String, Map<String, StatisticValue<?>>> listStatistics(
+      MetadataObject metadataObject, List<String> partitionNames) {
+    throw new UnsupportedOperationException(
+        "Don't support listStatistics with partition names yet.");
+  }
+
+  /**
+   * Appends statistics to the storage for a given metadata object.
+   *
+   * @param statisticsToAppend a map where the key is a {@link MetadataObject} and the value is a
+   *     {@link PartitionStatisticsUpdate} object
+   */
+  void appendStatistics(Map<MetadataObject, List<PartitionStatisticsUpdate>> statisticsToAppend);
+
+  /**
+   * Drops statistics for specified partitions of a metadata object.
+   *
+   * @param partitionStatisticsToDrop a map where the key is a {@link MetadataObject} and the value
+   *     is a list of {@link PartitionStatisticsDrop} objects representing the partitions and their
+   *     statistics to be dropped.
+   */
+  void dropStatistics(Map<MetadataObject, List<PartitionStatisticsDrop>> partitionStatisticsToDrop);
+
+  /**
+   * Updates statistics for a given metadata object. The default implementation is to first drop and
+   * then append the statistics. Developer can override this logic if needed.
+   *
+   * @param statisticsToUpdate a map where the key is a {@link MetadataObject} and the value is
+   *     another map containing partition names as keys and their corresponding statistics as
+   *     values.
+   */
+  default void updateStatistics(
+      Map<MetadataObject, List<PartitionStatisticsUpdate>> statisticsToUpdate) {
+    Map<MetadataObject, List<PartitionStatisticsDrop>> statisticsToDrop =
+        statisticsToUpdate.entrySet().stream()
+            .collect(
+                Collectors.toMap(
+                    Map.Entry::getKey,
+                    e ->
+                        e.getValue().stream()
+                            .map(
+                                entry ->
+                                    PartitionStatisticsDrop.of(
+                                        entry.partitionName(),
+                                        Lists.newArrayList(entry.statistics().keySet())))
+                            .collect(Collectors.toList())));
```

Comment:
This code is hard to review using stream. Also, how do you guarantee the thread safe thing?

Reply:
Add lock and add more data structure.